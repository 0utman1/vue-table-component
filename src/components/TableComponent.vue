<template>
    <div class="table-component">

        <div v-if="showFilter" class="table-component__filter">
            <input 
                class="table-component__filter__field" 
                type="text" 
                v-model="filter" 
                :placeholder="filterPlaceholder"
           >
            <a 
                v-if="filter"
                @click="filter = ''"
                class="table-component__filter__clear"
            >Ã—</a>
        </div>

        <div class="table-component__table-wrapper">
            <table :class="fullTableClass">
                <caption class="table-component__table__caption" role="alert" aria-live="polite">
                    {{ ariaCaption }}
                </caption>
                <thead>
                <tr>
                    <table-column-header
                        @click="changeSorting"
                        v-for="column in columns"
                        :key="column.properties.show"
                        :sort="sort"
                        :column="column"
                    />
                </tr>
                </thead>
                <tbody>
                <table-row
                    v-for="row in displayedRows"
                    :key="row.vueTableComponentInternalRowId"
                    :row="row"
                    :columns="columns"
                /></table-row>
                </tbody>
            </table>
        </div>

        <div v-if="displayedRows.length === 0" class="table-component__message">
            {{ filterNoResults }}
        </div>

        <div style="display:none;">
            <slot></slot>
        </div>
    </div>
</template>

<script>
    import Column from '../classes/Column';
    import expiringStorage from '../expiring-storage';
    import Row from '../classes/Row';
    import TableColumnHeader from './TableColumnHeader';
    import TableRow from './TableRow';
    import { pick } from 'lodash';
    import settings from '../settings';
    import { isArray, merge } from 'lodash';

    export default {
        components: {
            TableColumnHeader,
            TableRow,
        },

        props: {
            data: { required: true, type: Array },

            showFilter: { default: true },
            sortBy: { default: '', type: String },
            sortOrder: { default: 'desc', type: String },

            cacheId: { default: '' },
            cacheLifetime: { default: 5 },

            tableClass: { default: settings.tableClass },
            filterPlaceholder: { default: settings.filterPlaceholder },
            filterNoResults: { default: settings.filterNoResutls },
        },

        data: () => ({
            columns: [],
            rows: [],
            filter: '',
            sort: {
                fieldName: '',
                order: '',
            },
            localSettings: {},
        }),

        computed: {
            fullTableClass() {
                const extraClasses = isArray(this.tableClass) ? 
                    this.tableClass : 
                    [this.tableClass];

                return ['table-component__table', ...extraClasses];
            },

            ariaCaption() {
                if (this.sort.fieldName === '') {
                    return 'Table not sorted';
                }

                return `Table sorted by ${this.sort.fieldName} ` +
                    (this.sort.order === 'asc' ? '(ascending)' : '(descending)');
            },

            displayedRows() {
                return this.sortedRows.filter(row => row.passesFilter(this.filter));
            },

            sortedRows() {
                if (this.sort.fieldName === '') {
                    return this.rows;
                }

                if (this.columns.length === 0) {
                    return this.rows;
                }

                const sortColumn = this.getColumn(this.sort.fieldName);

                if (! sortColumn) {
                    return this.rows;
                }

                return this.rows.sort(sortColumn.getSortPredicate(this.sort.order, this.columns));
            },

            storageKey() {
                return `vue-table-component.${window.location.host}${window.location.pathname}${this.cacheId}`;
            },
        },

        watch: {
            filter() {
                this.saveState();
            },
        },

        mounted() {
            this.columns = this.$slots.default
                .filter(column => column.componentInstance)
                .map(column => pick(column.componentInstance, [
                    'show', 'label', 'dataType', 'sortable', 'sortOn', 'filterable', 'filterOn',
                ]))
                .map(columnProperties => new Column(columnProperties));

            let rowId = 0;

            this.rows = this.data
                .map(rowData => {
                    rowData.vueTableComponentInternalRowId = rowId++;
                    return rowData;
                })
                .map(rowData => new Row(rowData, this.columns));
        },

        created() {
            this.sort.fieldName = this.sortBy;
            this.sort.order = this.sortOrder;

            this.restoreState();
        },

        methods: {
            changeSorting(column) {
                if (this.sort.fieldName !== column.properties.show) {
                    this.sort.fieldName = column.properties.show;
                    this.sort.order = 'asc';
                }

                if (this.sort.fieldName === column.properties.show) {
                    this.sort.order = (this.sort.order === 'desc' ? 'asc' : 'desc');
                }

                this.saveState();
            },

            getColumn(columnName) {
                return this.columns.find(column => column.properties.show === columnName);
            },

            saveState() {
                expiringStorage.set(this.storageKey, pick(this.$data, ['filter', 'sort']), this.cacheLifetime);
            },

            restoreState() {
                const previousState = expiringStorage.get(this.storageKey);

                if (previousState === null) {
                    return;
                }

                this.filter = previousState.filter;
                this.sort = previousState.sort;

                this.saveState();
            },
        },
    };
</script>
