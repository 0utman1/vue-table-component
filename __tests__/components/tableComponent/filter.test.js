import TableComponent from '../../../src';
import Vue from 'vue/dist/vue.js';

describe('Filterable tableComponent', () => {
    Vue.use(TableComponent);

    beforeEach(() => {
        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        :data="[{ id: 1, firstName: 'John', lastName: 'Lennon' },
                                { id: 2, firstName: 'Paul', lastName: 'McCartney' },
                                { id: 3, firstName: 'George', lastName: 'Harrison' },
                                { id: 4, firstName: 'Ringo', lastName: 'Starr' }]"
                        sort-by="lastName"
                        sort-order="desc"
                    >
                        <table-column for="firstName" label="First name"></table-column>
                        <table-column for="lastName" label="Last name" :filterable="false"></table-column>
                    </table-component>
                </div>
            </div>
        `;


    });

    it('can filter data', async () => {

        const table = await createVm();

        table.filter = 'Paul';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will filter data in a case-insensitive way', async () => {

        const table = await createVm();

        table.filter = 'paul';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will display a message if there are no matching rows', async () => {

        const table = await createVm();

        table.filter = 'there are no rows that match this';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not use columns that are not filterable', async () => {

        const table = await createVm();

        table.filter = 'Lennon';

        await Vue.nextTick(() => {});

        expect(document.body.innerHTML).toMatchSnapshot();
    });


});



async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});

    const table = vm.$children[0];

    return table ;
}
