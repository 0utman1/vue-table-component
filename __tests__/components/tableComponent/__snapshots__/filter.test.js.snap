// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filterable tableComponent can filter data 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <a class=\\"table-component__filter__clear\\">×</a></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-desc\\">
    Last name
</th></tr></thead> <tbody><tr><td>Paul</td><td>McCartney</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableComponent can filter on another property 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <a class=\\"table-component__filter__clear\\">×</a></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th></tr></thead> <tbody><tr><td>Paul</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableComponent will display a message if there are no matching rows 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <a class=\\"table-component__filter__clear\\">×</a></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-desc\\">
    Last name
</th></tr></thead> <tbody></tbody></table></div> <div class=\\"table-component__message\\">
        There are no matching rows
    </div> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableComponent will filter data in a case-insensitive way 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <a class=\\"table-component__filter__clear\\">×</a></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-desc\\">
    Last name
</th></tr></thead> <tbody><tr><td>Paul</td><td>McCartney</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;

exports[`Filterable tableComponent will not use columns that are not filterable 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <a class=\\"table-component__filter__clear\\">×</a></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-desc\\">
    Last name
</th></tr></thead> <tbody></tbody></table></div> <div class=\\"table-component__message\\">
        There are no matching rows
    </div> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th></div></div></div></div>
        "
`;
